lesson-3-middleware-templating/practice/solutions.md

```diff
@@ -0,0 +1,72 @@
+## Lesson 3 - Practice
+
+0. In your own words, what is the purpose of middleware in web applications? Why is it useful?
+
+**Suggested Answer** In web applications, middleware allows different parts of the application to get access to a request and response object in different ways. This allows for easy filtering or common behavior to be abstracted between modules. This allows us to focus on writing code unique to our application while importing modules or writing small middleware pieces that can be used across applications.
+
+1. Write a custom middleware function for an express application that tracks how many times a page has been viewed and logs that count to the console.
+
+**Suggested Answer** Here's one way you might solve it.
+
+```js
+const app = express()
+
+const trackPageViews = (next) => {
```

Comment:
You use the parameter `req` in the body of this function but don't pass it through as a parameter. Shouldn't we  pass three parameters to this middleware function, with the first two representing the request and response? As in...
```js
const trackPageViews = (req, res, next) => {
  // ...
}

Reply:
Good catch - we're actually using a closure here, so `trackPageViews` shouldn't take any parameters. The function it returns is the actually middleware, which will take in the request, response, and next function.

Reply:
Oh, I see! I didn't see that it was a closure. So then yes, returning a function that takes three parameters makes sense. 

Thanks for making these changes on the fly! 🙏🏿